### ISSUE TITLE
dnn_mmod_ex training error on custom dataset

### ISSUE THREAD

**Issue by panovr** [davisking/dlib#578]
I use `dnn_mmod_ex` on a custom dataset, there is an error when traing:

> num training images: 7200
> num testing images:  2400
> detection window width,height:      42,38
> overlap NMS IOU thresh:             0
> overlap NMS percent covered thresh: 0
> step#: 0     learning rate: 0.1   average loss: 0            steps without apparent progress: 0
> Encountered a truth rectangle with a width and height of 17 and 45.
> The image pyramid and sliding window can't output a rectangle of this shape. 
> This is because the rectangle is smaller than the detection window which has a width
> and height of 42 and 38.

I checked and searched in the `training.xml` file, and can't find any box with width 17 and height 45.
May I ask that how can I debug this error and find which image file cause this error?
Thanks!

**Comment by davisking** [comment]
That example crops out random parts of your input images and those images are what go into the training code.  It never sees your original input images.  So you have to look at the images *going into the DNN training* to see what is happening.

**Comment by panovr** [comment]
You are right, and I found there are some errors in my bounding boxes annotations when I convert them from JSON to XML format. I `use random_cropper_ex` to validate this error.
Thanks for pointing out this and thanks for sharing dlib!

**Comment by panovr** [comment]
My random_cropper settings are:
```
random_cropper cropper;
cropper.set_chip_dims(200, 200);
cropper.set_min_object_height(0.2);
```
I use `random_cropper_ex` to visualize some cropped samples:
- regular image with object bounding box
![i1](https://4.bp.blogspot.com/-kSQESszNjIM/WQ9F8US5JTI/AAAAAAAAA_w/KClHUgxxoiw3-W-4MwCKwFVFp2xo5oJmwCLcB/s1600/i1.png)
- only part of image and part of bounding box
![i2](https://1.bp.blogspot.com/-RUxcDkmQIJQ/WQ9F8tljApI/AAAAAAAAA_0/gp_MmHxmJT0BJTUFlDBoXbkmFAtYqZH6wCLcB/s1600/i2.png)
- background
![i3](https://3.bp.blogspot.com/-5UMpHNwPZrk/WQ9F8W477wI/AAAAAAAAA_4/c-0Ytj_kF7MC3gH2ICgh3lfWl1gZXIM7QCLcB/s1600/i3.png)

@davisking from the second image and the third image, you can see the bounding boxes are out of bounds of the cropped images. Will this be a problem to `dnn_mmod_ex` training?

**Comment by davisking** [comment]
It's only a problem if it isn't set to ignore.

**Comment by panovr** [comment]
```
num training images: 7200
num testing images:  2400
detection window width,height:      21,19
overlap NMS IOU thresh:             0
overlap NMS percent covered thresh: 0
step#: 0     learning rate: 0.1   average loss: 0            steps without apparent progress: 0
Encountered a truth rectangle with a width and height of 159 and 60.
The image pyramid and sliding window can't output a rectangle of this shape. 
This is because the rectangle's aspect ratio is too different from the detection window,
which has a width and height of 21 and 19.
```
May I ask that how can I adjust the detection window dimension to fit the ground truth rectangle dimension?

**Comment by davisking** [comment]
You have to make your truth boxes more uniform in aspect ratio.

**Comment by panovr** [comment]
Face annotations are more uniform in aspect ratio than my custom data annotations. Thanks for pointing out this.
